homeassistant/components/websocket_api/commands.py

```diff
@@ -242,33 +252,43 @@ def handle_get_states(
     """Handle get states command."""
     states = _async_get_allowed_states(hass, connection)
 
-    # JSON serialize here so we can recover if it blows up due to the
-    # state machine containing unserializable data. This command is required
-    # to succeed for the UI to show.
-    response = messages.result_message(msg["id"], states)
     try:
-        connection.send_message(JSON_DUMP(response))
-        return
+        serialized_states = [state.as_dict_json() for state in states]
     except (ValueError, TypeError):
-        connection.logger.error(
-            "Unable to serialize to JSON. Bad data found at %s",
-            format_unserializable_data(
-                find_paths_unserializable_data(response, dump=JSON_DUMP)
-            ),
-        )
-    del response
+        pass
+    else:
+        _send_handle_get_states_response(connection, msg["id"], serialized_states)
+        return
 
     # If we can't serialize, we'll filter out unserializable states
-    serialized = []
+    serialized_states = []
     for state in states:
-        # Error is already logged above
-        with suppress(ValueError, TypeError):
-            serialized.append(JSON_DUMP(state))
+        try:
+            serialized_states.append(state.as_dict_json())
+        except (ValueError, TypeError):
+            connection.logger.error(
+                "Unable to serialize to JSON. Bad data found at %s",
+                format_unserializable_data(
+                    find_paths_unserializable_data(state, dump=JSON_DUMP)
+                ),
+            )
+
+    _send_handle_get_states_response(connection, msg["id"], serialized_states)
 
-    # We now have partially serialized states. Craft some JSON.
-    response2 = JSON_DUMP(messages.result_message(msg["id"], ["TO_REPLACE"]))
-    response2 = response2.replace('"TO_REPLACE"', ", ".join(serialized))
-    connection.send_message(response2)
+
+def _send_handle_get_states_response(
+    connection: ActiveConnection, msg_id: int, serialized_states: list[str]
+) -> None:
+    """Send handle get states response."""
+    connection.send_message(
+        _HANDLE_GET_STATES_TEMPLATE.replace(
```

Comment:
In a future PR I think we can make this a bit more efficient by making a function and can construct event and result messages without replace. I'm going to explore that and if successful will do a PR that does that for every place we do this.  I am not 100% sure it will work, just an idea at this point